core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -2310,9 +2376,9 @@ public void onMatch(RelOptRuleCall call) {
                   Pair.of(rexBuilder.makeLiteral(true),
                       "nullIndicator")));
 
-      LogicalJoin join =
-          LogicalJoin.create(left, right, joinCond,
-              ImmutableSet.of(), joinType);
+      Join join =
+          (Join) relBuilder.push(left).push(right)
```

Comment:
is it guaranteed to return a join?

Reply:
We are not calling the function with correlated fields, and, if that is the case, this join builder always seems to call createJoin method from factory. Alternatively, I think we could check the type of it and only call projectJoinOutputWithNullability function if it is join instance.